File: go/mpp/adapters/mpp_adapter.go
package adapters

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"

	"github.com/okx/payments/go/mpp/protocol"
	"github.com/okx/payments/go/mpp/server"
	"github.com/okx/payments/go/paymentrouter"
)

// Ensure MppAdapter implements the ProtocolAdapter interface.
var _ paymentrouter.ProtocolAdapter = (*MppAdapter)(nil)

// MppRouteConfig holds per-route configuration for the MppAdapter.
type MppRouteConfig struct {
	Intent           string `json:"intent"`
	Amount           string `json:"amount"`
	Currency         string `json:"currency"`
	Decimals         uint32 `json:"decimals"`
	Description      string `json:"description,omitempty"`
	ExternalID       string `json:"externalId,omitempty"`
	Realm            string `json:"realm,omitempty"`
	UnitType         string `json:"unitType,omitempty"`
	SuggestedDeposit string `json:"suggestedDeposit,omitempty"`
}

// MppAdapter implements paymentrouter.ProtocolAdapter for the MPP payment protocol.
type MppAdapter struct {
	mpp      *server.Mpp
	priority int
}

// NewMppAdapter creates a new MppAdapter wrapping the provided Mpp server instance.
func NewMppAdapter(mpp *server.Mpp) *MppAdapter {
	return &MppAdapter{mpp: mpp, priority: 10}
}

// Name returns the protocol name identifier.
func (a *MppAdapter) Name() string {
	return "mpp"
}

// Priority returns the adapter's priority for protocol selection ordering.
func (a *MppAdapter) Priority() int {
	return a.priority
}

// Detect reports whether the request carries a PaymentAuth Authorization header.
func (a *MppAdapter) Detect(r *http.Request) bool {
	authHeader := r.Header.Get(protocol.AuthorizationHeader)
	if authHeader == "" {
		return false
	}
	idx := strings.IndexByte(authHeader, ' ')
	var scheme string
	if idx < 0 {
		scheme = authHeader
	} else {
		scheme = authHeader[:idx]
	}
	return scheme == protocol.PaymentScheme
}

// GetChallenge issues a WWW-Authenticate challenge header based on the route config.
// Returns nil headers (without error) when cfg is nil or has no intent set.
func (a *MppAdapter) GetChallenge(ctx context.Context, r *http.Request, cfg any) (http.Header, error) {
	if cfg == nil {
		return nil, nil
	}

	routeCfg, ok := cfg.(MppRouteConfig)
	if !ok {
		return nil, fmt.Errorf("mpp adapter: invalid config type %T", cfg)
	}

	if routeCfg.Intent == "" {
		return nil, nil
	}

	var challengeHeader string
	var err error

	switch routeCfg.Intent {
	case protocol.IntentCharge:
		challengeHeader, err = a.mpp.Charge(ctx, server.ChargeRouteConfig{
			Amount:      routeCfg.Amount,
			Currency:    routeCfg.Currency,
			Decimals:    routeCfg.Decimals,
			Description: routeCfg.Description,
			ExternalID:  routeCfg.ExternalID,
		})
	case protocol.IntentSession:
		challengeHeader, err = a.mpp.SessionChallenge(ctx, server.SessionRouteConfig{
			Amount:           routeCfg.Amount,
			Currency:         routeCfg.Currency,
			Decimals:         routeCfg.Decimals,
			Description:      routeCfg.Description,
			ExternalID:       routeCfg.ExternalID,
			UnitType:         routeCfg.UnitType,
			SuggestedDeposit: routeCfg.SuggestedDeposit,
		})
	default:
		return nil, fmt.Errorf("mpp: unknown intent %q", routeCfg.Intent)
	}

	if err != nil {
		return nil, fmt.Errorf("mpp: generate challenge: %w", err)
	}

	h := make(http.Header)
	h.Set(protocol.WWWAuthenticateHeader, challengeHeader)
	return h, nil
}

// Handle verifies the payment credential in the Authorization header and writes
// the Payment-Receipt header on success. Returns an error on failure after
// writing an error JSON body to w.
func (a *MppAdapter) Handle(w http.ResponseWriter, r *http.Request, cfg any) error {
	authHeader := r.Header.Get(protocol.AuthorizationHeader)
	if authHeader == "" {
		writeErrorJSON(w, http.StatusUnauthorized, "missing Authorization header")
		return fmt.Errorf("mpp: missing Authorization header")
	}

	challengeHeader, err := server.ChallengeHeaderFromAuth(authHeader)
	if err != nil {
		writeErrorJSON(w, http.StatusBadRequest, "invalid Authorization header")
		return fmt.Errorf("mpp: reconstruct challenge: %w", err)
	}

	// Determine intent from the reconstructed challenge.
	challenge, err := protocol.PaymentChallengeFromHeader(challengeHeader)
	if err != nil {
		writeErrorJSON(w, http.StatusBadRequest, "invalid challenge header")
		return fmt.Errorf("mpp: parse challenge: %w", err)
	}

	switch challenge.Intent {
	case protocol.IntentCharge:
		receipt, err := a.mpp.VerifyCredential(r.Context(), challengeHeader, authHeader)
		if err != nil {
			writeErrorJSON(w, http.StatusUnauthorized, "payment verification failed")
			return fmt.Errorf("mpp: verify charge: %w", err)
		}
		if receiptHeader, hErr := receipt.ToHeader(); hErr == nil {
			w.Header().Set(protocol.PaymentReceiptHeader, receiptHeader)
		}

	case protocol.IntentSession:
		result, err := a.mpp.VerifySession(r.Context(), challengeHeader, authHeader)
		if err != nil {
			writeErrorJSON(w, http.StatusUnauthorized, "payment verification failed")
			return fmt.Errorf("mpp: verify session: %w", err)
		}
		if receiptHeader, hErr := result.Receipt.ToHeader(); hErr == nil {
			w.Header().Set(protocol.PaymentReceiptHeader, receiptHeader)
		}
		if result.ManagementResponse != nil {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusOK)
			json.NewEncoder(w).Encode(result.ManagementResponse) //nolint:errcheck
			return nil
		}

	default:
		writeErrorJSON(w, http.StatusBadRequest, fmt.Sprintf("unknown intent %q", challenge.Intent))
		return fmt.Errorf("mpp: unknown intent %q", challenge.Intent)
	}

	return nil
}

// writeErrorJSON writes a JSON error body with the given status code.
func writeErrorJSON(w http.ResponseWriter, status int, msg string) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(map[string]string{"error": msg}) //nolint:errcheck
}

File: python/mpp/src/mpp_evm/adapters/mpp_adapter.py
"""PaymentRouter ProtocolAdapter for MPP.

Wraps an Mpp server instance and delegates detect/challenge/handle to pympp.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from mpp.server.mpp import Mpp


@dataclass
class MppRouteConfig:
    """Per-route configuration for MppAdapter.

    """

    intent: str = ""
    amount: str = ""
    currency: str = ""
    decimals: int = 6
    description: str = ""
    external_id: str = ""
    realm: str = ""
    unit_type: str = ""
    suggested_deposit: str = ""


class MppAdapter:
    """PaymentRouter ProtocolAdapter for MPP protocol.

    Wraps an Mpp server instance. Detects requests with
    'Authorization: Payment ...' header.
    """

    def __init__(self, mpp: Mpp, priority: int = 10) -> None:
        self._mpp = mpp
        self._priority = priority

    @property
    def name(self) -> str:
        return "mpp"

    @property
    def priority(self) -> int:
        return self._priority

    def detect(self, request: Any) -> bool:
        auth = request.get_header("authorization")
        if not auth:
            return False
        scheme = auth.split(" ", 1)[0]
        return scheme == "Payment"

    async def get_challenge(self, request: Any, cfg: Any) -> dict[str, str]:
        if cfg is None:
            return {}

        route_cfg = cfg if isinstance(cfg, MppRouteConfig) else MppRouteConfig(**cfg)
        if not route_cfg.intent:
            return {}

        intent_obj = self._mpp.method.intents.get(route_cfg.intent)
        if intent_obj is None:
            return {}

        from mpp.server.verify import verify_or_challenge
        from mpp.server.decorator import make_challenge_response
        from mpp import Challenge
        DEFAULT_DECIMALS = 6
        from mpp._units import parse_units

        decimals = getattr(self._mpp.method, "decimals", DEFAULT_DECIMALS)
        base_amount = str(parse_units(route_cfg.amount, decimals))
        currency = route_cfg.currency or getattr(self._mpp.method, "currency", "")
        recipient = getattr(self._mpp.method, "recipient", "")

        req: dict[str, Any] = {
            "amount": base_amount,
            "currency": currency,
            "recipient": recipient,
        }
        if route_cfg.description:
            req["description"] = route_cfg.description
        if route_cfg.unit_type:
            req["unitType"] = route_cfg.unit_type
        if route_cfg.suggested_deposit:
            req["suggestedDeposit"] = route_cfg.suggested_deposit
        if hasattr(intent_obj, "challenge_method_details"):
            req["methodDetails"] = intent_obj.challenge_method_details()

        result = await verify_or_challenge(
            authorization=None,
            intent=intent_obj,
            request=req,
            realm=self._mpp.realm,
            secret_key=self._mpp.secret_key,
            method=self._mpp.method.name,
            description=route_cfg.description or None,
        )

        if isinstance(result, Challenge):
            # Format as WWW-Authenticate header
            resp = make_challenge_response(result, self._mpp.realm)
            www_auth = resp.headers.get("www-authenticate", "")
            if www_auth:
                return {"WWW-Authenticate": www_auth}
        return {}

    async def handle(self, request: Any, cfg: Any) -> Any:
        """Verify payment credential from Authorization header.

        Delegates to Mpp's verify flow. For charge intent, returns the receipt.
        For session intent, returns the SessionVerifyResult (receipt + management response).
        """
        from mpp.server.verify import verify_or_challenge
        from mpp import Challenge

        route_cfg = cfg if isinstance(cfg, MppRouteConfig) else MppRouteConfig(**cfg)

        auth_header = request.get_header("authorization")
        if not auth_header:
            raise ValueError("missing Authorization header")

        intent_obj = self._mpp.method.intents.get(route_cfg.intent)
        if intent_obj is None:
            raise ValueError(f"unknown intent: {route_cfg.intent}")

        # Build request params matching what the decorator would build
        DEFAULT_DECIMALS = 6
        from mpp._units import parse_units

        decimals = getattr(self._mpp.method, "decimals", DEFAULT_DECIMALS)
        base_amount = str(parse_units(route_cfg.amount, decimals))
        currency = route_cfg.currency or getattr(self._mpp.method, "currency", "")
        recipient = getattr(self._mpp.method, "recipient", "")

        req: dict[str, Any] = {
            "amount": base_amount,
            "currency": currency,
            "recipient": recipient,
        }

        if route_cfg.description:
            req["description"] = route_cfg.description
        if route_cfg.unit_type:
            req["unitType"] = route_cfg.unit_type
        if route_cfg.suggested_deposit:
            req["suggestedDeposit"] = route_cfg.suggested_deposit

        if hasattr(intent_obj, "challenge_method_details"):
            req["methodDetails"] = intent_obj.challenge_method_details()

        result = await verify_or_challenge(
            authorization=auth_header,
            intent=intent_obj,
            request=req,
            realm=self._mpp.realm,
            secret_key=self._mpp.secret_key,
            method=self._mpp.method.name,
            description=route_cfg.description or None,
        )

        if isinstance(result, Challenge):
            # Should not happen since detect() already confirmed payment present
            raise ValueError("unexpected challenge during handle")

        credential, receipt = result

        # For session: check respond() for management actions
        respond_fn = getattr(intent_obj, "respond", None)
        if respond_fn is not None:
            mgmt = respond_fn(credential, receipt)
            if mgmt is not None:
                return {"_management": True, "response": mgmt}

        return receipt
